File: detection/semgrep/fixtures/batch-resolve-missing-per-item-scope-filter.js
// Test fixture for rule: mcp-batch-resolve-missing-per-item-scope-filter
//
// S12 (batch/bulk endpoint BOLA). A batch endpoint resolves multiple ids
// but skips the per-item org check. The sibling list() tool is correctly scoped
// while the batch-get sibling skips the per-object check — the "list->get
// asymmetry" pattern at batch scale.
//
// Two-way canary axis: is there a per-item filter back to the caller's scope?
// (present ⇒ SILENT)

async function batchResolveVuln({ ids }) {
  // ruleid: mcp-batch-resolve-missing-per-item-scope-filter
  return ids.map(id => store.getNote(id)).filter(Boolean);
}

async function batchResolveSafe({ ids, session }) {
  // ok: mcp-batch-resolve-missing-per-item-scope-filter
  const resolved = ids.map(id => store.getNote(id)).filter(Boolean);
  return resolved.filter(n => n.orgId === session.orgId);
}

async function batchResolveWithFilter({ ids, session }) {
  // ok: mcp-batch-resolve-missing-per-item-scope-filter
  const resolved = ids.map(id => store.getNote(id)).filter(Boolean);
  return resolved.filter(n => n.orgId === session.orgId);
}

File: detection/semgrep/fixtures/batch-resolve-missing-per-item-scope-filter.py
# Test fixture for rule: mcp-batch-resolve-missing-per-item-scope-filter-py
#
# S12 (batch/bulk endpoint BOLA). Python spelling: list comprehension
# resolving multiple ids without per-item scope filter.

def batch_resolve_vuln(ids):
    # ruleid: mcp-batch-resolve-missing-per-item-scope-filter-py
    return [store.get_note(f) for f in ids]

def batch_resolve_safe(ids, session):
    # ok: mcp-batch-resolve-missing-per-item-scope-filter-py
    resolved = [store.get_note(f) for f in ids]
    return [n for n in resolved if n.org_id == session.org_id]

def batch_resolve_with_filter(ids, session):
    # ok: mcp-batch-resolve-missing-per-item-scope-filter-py
    resolved = [store.get_note(f) for f in ids]
    return [n for n in resolved if n.org_id == session.org_id]